### PR TITLE
Loan API Enhancements

### LINKED ISSUE #63: Loan API Enhancements
Create a loan in a folder
Create a loan from a loan template
Apply a loan template to an existing loan
Move a loan from one folder to another
Retrieve metadata information from a loan

### PR DESCRIPTION
**Pull request by TylerBrinkley**
Create a loan in a folder
Create a loan from a loan template
Apply a loan template to an existing loan
Move a loan from one folder to another
Retrieve metadata information from a loan

Closes #63

Still needs tests added

### CODE REVIEW

File: src/EncompassRest/ApiObject.cs [hunk 1]
```diff
@@ -64,21 +64,21 @@ internal ApiObject(EncompassRestClient client, string baseApiPath)
...
+        internal Task PutPopulateDirtyAsync(string requestUri, string queryString, HttpContent content, string methodName, string resourceId, CancellationToken cancellationToken, IDirty target, bool populate) => PopulateDirtyInternalAsync(HttpMethod.Put, requestUri, queryString, content, methodName, resourceId, cancellationToken, target, populate);
 
         internal Task<T> PutAsync<T>(string requestUri, string queryString, HttpContent content, string methodName, string resourceId, CancellationToken cancellationToken, Func<HttpResponseMessage, Task<T>> func) => SendAsync(HttpMethod.Put, requestUri, queryString, content, methodName, resourceId, cancellationToken, func);
 
         internal Task<bool> DeleteAsync(string requestUri, CancellationToken cancellationToken) => SendAsync(HttpMethod.Delete, requestUri, null, null, null, null, cancellationToken, IsSuccessStatusCodeFunc, false);
```
Comment (hfcjweinstock) [inline src/EncompassRest/ApiObject.cs]: delete needs query params too

File: src/EncompassRest/Loans/UpdateLoanOptions.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
+﻿using EncompassRest.Utilities;
+
+namespace EncompassRest.Loans
+{
+    public sealed class UpdateLoanOptions : LoanOptions
+    {
+        public bool AppendData { get; set; }
```
Comment (hfcjweinstock) [inline src/EncompassRest/Loans/UpdateLoanOptions.cs]: we should have defaults for this.

File: src/EncompassRest/Loans/LoanOptions.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
+﻿using EncompassRest.Utilities;
+
+namespace EncompassRest.Loans
+{
+    public abstract class LoanOptions
+    {
+        public bool Populate { get; set; }
```
Comment (hfcjweinstock) [inline src/EncompassRest/Loans/LoanOptions.cs]: we should have defaults for this. - false

File: src/EncompassRest/Loans/LoanOptions.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        internal LoanOptions()
+        {
+        }
+
+        internal virtual QueryParameters ToQueryParameters()
+        {
+            var queryParameters = new QueryParameters();
+            if (Populate)
+            {
+                queryParameters.Add("view", "entity");
```
Comment (hfcjweinstock) [inline src/EncompassRest/Loans/LoanOptions.cs]: queryParameters.Add("view", populate?"entity":"id");

File: src/EncompassRest/Loans/LoanApiObject.cs [hunk 5]
```diff
@@ -67,7 +67,7 @@ internal LoanApiObject(EncompassRestClient client, string loanId, string baseApi
         internal async Task<string> CreateAsync<TClass>(TClass value, string methodName, bool populate, CancellationToken cancellationToken)
             where TClass : class, T, IDirty, IIdentifiable
         {
-            var id = await PostPopulateDirtyAsync(null, value, methodName, populate, cancellationToken).ConfigureAwait(false);
+            var id = await PostPopulateDirtyAsync(null, populate ? ViewEntityQueryString : null, value, methodName, populate, cancellationToken).ConfigureAwait(false);
```
Comment (hfcjweinstock) [inline src/EncompassRest/Loans/LoanApiObject.cs]: we need to allow for more querystring options somehow.

File: src/EncompassRest/Loans/LoanApiObject.cs [hunk 6]
```diff
@@ -78,7 +78,7 @@ internal LoanApiObject(EncompassRestClient client, string loanId, string baseApi
         internal async Task UpdateAsync<TClass>(TClass value, string methodName, bool populate, CancellationToken cancellationToken)
             where TClass : class, T, IDirty, IIdentifiable
         {
-            await PatchPopulateDirtyAsync(value.Id, JsonStreamContent.Create(value), methodName, value.Id, cancellationToken, value, populate).ConfigureAwait(false);
+            await PatchPopulateDirtyAsync(value.Id, populate ? ViewEntityQueryString : null, JsonStreamContent.Create(value), methodName, value.Id, cancellationToken, value, populate).ConfigureAwait(false);
```
Comment (hfcjweinstock) [inline src/EncompassRest/Loans/LoanApiObject.cs]: we need to allow for more querystring options somehow.

File: src/EncompassRest/ApiObject.cs (hunk 1 above)
Comment (TylerBrinkley) [reply src/EncompassRest/ApiObject.cs]: Probably but it wasn't required for this particular change. I can do it in this PR though.

File: src/EncompassRest/Loans/LoanApiObject.cs (hunk 5 above)
Comment (TylerBrinkley) [reply src/EncompassRest/Loans/LoanApiObject.cs]: Agreed

File: src/EncompassRest/Loans/LoanApiObject.cs (hunk 6 above)
Comment (TylerBrinkley) [reply src/EncompassRest/Loans/LoanApiObject.cs]: Agreed

File: src/EncompassRest/Loans/LoanOptions.cs (hunk 3 above)
Comment (TylerBrinkley) [reply src/EncompassRest/Loans/LoanOptions.cs]: If you don't assign a value to it the default is `false`.

File: src/EncompassRest/Loans/LoanOptions.cs (hunk 4 above)
Comment (TylerBrinkley) [reply src/EncompassRest/Loans/LoanOptions.cs]: We don't need "id" as we'll get the id from the location header.

File: src/EncompassRest/Loans/UpdateLoanOptions.cs (hunk 2 above)
Comment (TylerBrinkley) [reply src/EncompassRest/Loans/UpdateLoanOptions.cs]: Agreed, I think this one might ought to be `true` so as not to overwrite existing data. Thoughts?

File: src/EncompassRest/Loans/LoanOptions.cs (hunk 4 above)
Comment (hfcjweinstock) [reply src/EncompassRest/Loans/LoanOptions.cs]: the 2 options for view are entity or id.  should we pass it always or only when not default?

File: src/EncompassRest/Loans/LoanOptions.cs (hunk 3 above)
Comment (hfcjweinstock) [reply src/EncompassRest/Loans/LoanOptions.cs]: ok

File: src/EncompassRest/ApiObject.cs (hunk 1 above)
Comment (hfcjweinstock) [reply src/EncompassRest/ApiObject.cs]: ok.  i need it for the resourceLocking.

File: src/EncompassRest/Loans/LoanOptions.cs (hunk 3 above)
Comment (TylerBrinkley) [reply src/EncompassRest/Loans/LoanOptions.cs]: I could make it explicit though I suppose.

File: src/EncompassRest/Loans/LoanApiObject.cs (hunk 5 above)
Comment (TylerBrinkley) [reply src/EncompassRest/Loans/LoanApiObject.cs]: Actually, I don't think that's necessary and would just make these API's harder to implement. I think if we ever come across an instance in the future where we need to allow for more querystring options we can just add overloads.

File: src/EncompassRest/Loans/LoanApiObject.cs (hunk 6 above)
Comment (TylerBrinkley) [reply src/EncompassRest/Loans/LoanApiObject.cs]: Actually, I don't think that's necessary and would just make these API's harder to implement. I think if we ever come across an instance in the future where we need to allow for more querystring options we can just add overloads.

File: src/EncompassRest/Loans/UpdateLoanOptions.cs (hunk 2 above)
Comment (hfcjweinstock) [reply src/EncompassRest/Loans/UpdateLoanOptions.cs]: i would mirror encompass default functionality or make sure we have the documentation to let people know what the default is and does

File: src/EncompassRest/Loans/LoanOptions.cs (hunk 4 above)
Comment (TylerBrinkley) [reply src/EncompassRest/Loans/LoanOptions.cs]: id is no use to us as it's included in the header thus I think we should only support entity when using populate and don't include a view query parameter at all if not populate.

File: src/EncompassRest/Loans/LoanApiObject.cs (hunk 6 above)
Comment (hfcjweinstock) [reply src/EncompassRest/Loans/LoanApiObject.cs]: ok.  i was just thinking about future proofing.

File: src/EncompassRest/Loans/LoanApiObject.cs (hunk 5 above)
Comment (hfcjweinstock) [reply src/EncompassRest/Loans/LoanApiObject.cs]: ok.  i was just thinking about future proofing.

File: src/EncompassRest/Loans/UpdateLoanOptions.cs (hunk 2 above)
Comment (TylerBrinkley) [reply src/EncompassRest/Loans/UpdateLoanOptions.cs]: The question is what is Encompass's default functionality.

Comment (hfcjweinstock) [reply src/EncompassRest/Loans/UpdateLoanOptions.cs]: default is false.

Comment (TylerBrinkley) [reply src/EncompassRest/Loans/UpdateLoanOptions.cs]: Interesting, thanks for figuring that out.

File: src/EncompassRest/Loans/LoanApiObject.cs (hunk 5 above)
Comment (TylerBrinkley) [reply src/EncompassRest/Loans/LoanApiObject.cs]: I understand. We'll cross that bridge when we get there.

File: src/EncompassRest/Loans/LoanOptions.cs (hunk 4 above)
Comment (hfcjweinstock) [reply src/EncompassRest/Loans/LoanOptions.cs]: ok

**Pull request merged**